rfcs/2021-10-29-8621-framing-and-codecs-sinks.md

```diff
@@ -0,0 +1,126 @@
+- Strategy for integrating encoding/framing with sinks
+
+### Out of scope
+
+- Integration with schema system (automatic transformation of event fields depending)
+
+## Pain
+
+Internally, we don't share implementations for encoding strategies that are very commonly used (e.g. JSON, newline-delimited). This leads to duplicated effort and missing guidance for handling encoding in sinks since no canonical implementation exists.
+
+Externally, users are restricted by the available custom set of encodings implemented by the sink. There does not exist any mental model for consistent encoding/framing options for event data that applies to all sinks.
+
+## Proposal
+
+### User Experience
+
+Users should be able to set `encoding` and `framing` options on sinks, analogously to the `decoding` and `framing` options on sources. These options uniformly control how the event _payload_ is encoded. This distinction is important, as encoding for the sink specific _protocol_ and the event _payload_ are separate concerns. The payload should still be encoded according to the sink's protocol, and the sink should provide additional options if there are multiple protocols to choose from, e.g. under a `protocol` key.
+
+The fields containing event transformations in sinks on the current `encoding` options (`schema`, `only_fields`, `except_fields`, `timestamp_format`) should be moved to a dedicated option, e.g. `schema` or `transform`. Thus, this would introduce a breaking change in configurations. However, migration would be relatively straight-forward by nesting these options under the new key.
+
+### Implementation
+
+The following common config structures will be provided, analogously to the decoding/framing configs:
+
+```rust
+pub trait SerializerConfig: Debug + DynClone + Send + Sync {
+    /// Builds a serializer from this configuration.
+    ///
+    /// Fails if the configuration is invalid.
+    fn build(&self) -> crate::Result<BoxedSerializer>;
+}
+```
+
+```rust
+pub trait FramingConfig: Debug + DynClone + Send + Sync {
+    /// Builds a framer from this configuration.
+    ///
+    /// Fails if the configuration is invalid.
+    fn build(&self) -> crate::Result<BoxedFramer>;
+}
+```
+
+These can be build to form an `Encoder`:
+
+```rust
+/// An encoder that can encode structured events to byte messages.
+pub struct Encoder {
+    serializer: BoxedSerializer,
+    framer: BoxedFramer,
+}
+```
+
+`Encoder` implements `tokio_util::codec::Encoder<SmallVec<[Event; 1]>>`. Internally, events first go through the `Serializer` which implements `tokio_util::codec::Encoder<SmallVec<[Event; 1]>>` and are then handed over to the `Framer` which implements `tokio_util::codec::Encoder<Bytes>`.
+
+## Plan Of Attack
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is approved:
+
+- [ ] Implementation of support structures (`Encoder`, `SerializerConfig`, `FramingConfig`)
```

Comment:
Probably worth mentioning some kind of backwards-compat layer here so that we can transition the actual user-facing config API on a separate cadence from the implementation.

Reply:
Added a note in 7e087fc5a526bc13d1511a959af7263efd866d46